runtime/doc/lsp.txt

```diff
@@ -954,6 +954,20 @@ start_client({config})                                *vim.lsp.start_client()*
                     not be fully initialized. Use `on_init` to do any actions
                     once the client has been initialized.
 
+start_or_attach({config}, {reuse_client})          *vim.lsp.start_or_attach()*
+                Starts a new client or re-uses an existing client and attaches
+                the current buffer to it.
+
+                Parameters: ~
+                    {config}        (table) Same configuration as documented
+                                    in |lsp.start_client()|
+                    {reuse_client}  nil|fun(client: table): boolean Function
```

Comment:
so reuse_client=false is literally the same thing as start_client? we don't need a new function for this

Reply:
It also attaches the current buffer. We'd probably have to add two parameters to `start_client`, `reuse_client` and `(auto)attach = true|false` to keep the existing behavior by default. And then users would have to add an explicit `attach = true` to their configuration to opt into the new behavior. I'm not sure that would be an improvement over this.

Reply:
ok, your call.  I wonder though, if we had this feature from the beginning, would we make multiple functions for it or would it be a feature of start_client() ?

Reply:
> I wonder though, if we had this feature from the beginning, would we make multiple functions for it or would it be a feature of start_client() ?

Probably not. It still might be better to add a new function that does everything right, and then deprecate the old functions. (The biggest consumer of the old functions is nvim-lspconfig, which will be made obsolete -- as far as code goes -- by this PR anyway.)

Reply:
> I wonder though, if we had this feature from the beginning, would we make multiple functions for it or would it be a feature of start_client() ?

Probably a single function.
I gave this some more thought and came to the same conclusion as @clason